--- a/arch/arm64/boot/dts/amlogic/meson-gxbb-wetek.dtsi	2018-06-06 08:09:26.970866789 +0100
+++ b/arch/arm64/boot/dts/amlogic/meson-gxbb-wetek.dtsi	2018-06-06 08:12:17.642339887 +0100
@@ -108,13 +108,13 @@
 
 	sound {
 		compatible = "simple-audio-card";
-		simple-audio-card,name = "meson-gx-preview";
+		simple-audio-card,name = "HDMI";
 		status = "okay";
 
 		simple-audio-card,dai-link@0 {
 			/* HDMI Output */
 			format = "i2s";
-			mclk-fs = <256>;
+			mclk-fs = <128>;
 			bitclock-master =  <&i2s_dai>;
 			frame-master = <&i2s_dai>;
 			plat {
